driver-core/src/main/com/mongodb/internal/session/ServerSessionPool.java

```diff
@@ -232,32 +223,4 @@ public boolean isMarkedDirty() {
             return dirty;
         }
     }
-
-    private final class ServerSessionItemFactory implements ConcurrentPool.ItemFactory<ServerSessionImpl> {
-        @Override
-        public ServerSessionImpl create() {
-            return new ServerSessionImpl(createNewServerSessionIdentifier());
-        }
-
-        @Override
-        public void close(final ServerSessionImpl serverSession) {
-            closeSession(serverSession);
```

Comment:
Previously, we were requesting the server to release its resources for closed sessions (not on every `close`, only when we reach `END_SESSIONS_BATCH_SIZE`, but nonetheless). Now we don't do that at all while the pool is used, and only ask the server to release resources for those session that are still in the pool when the pool is being closed. I don't think that this is really what we want, unless I am misunderstanding the situation.

Unless we are going to get rid of `endClosedSessions`, we also need to make sure that sessions released to the pool concurrently with it being closed do not stay unclosed, and that the server also gets notified about them. The version of the pool before the PR does this, but I have not tried to assess the correctness.

Reply:
The previous code was very confusing and you may be misinterpreting it.  Note that `closeSession` was a no-op unless the pool itself was in the process of being closed (see the comment `only track closed sessions when pool is in the process of closing`).

Reply:
True, I confused myself. Thanks to explaining to me what is done and why over Zoom.